### PR TITLE
Fallback to quotation mark if the email is too big to parse

### PR DESCRIPTION
**Pull request by evonck**
If the message is too bug try to rely on quotation mark to find the quote.
This will prevent us from having too many untouched email.
We might miss some added quote in big message.

### CODE REVIEW

File: src/Quotations.ts [hunk 1]
```diff
@@ -111,49 +105,66 @@ export function extractFromHtml(messageBody: string, options?: ExtractFromHtmlOp
   if (xmlDocument.lastChild && xmlDocument.lastChild.nodeValue)
     xmlDocument.removeChild(xmlDocument.lastChild);
 
-  // Keep a copy of the original document around.
+  // Add the checkpoints to the HTML tree.
+  const result = extractQuotationFromCheckpoint(xmlDocument, messageBody, options);
```
Comment (evonck) [inline src/Quotations.ts]: Naming is hard I did not find anything better.

File: src/Quotations.ts [hunk 2]
```diff
@@ -111,49 +105,66 @@ export function extractFromHtml(messageBody: string, options?: ExtractFromHtmlOp
...
+  if (result.didFindQuote && !result.isTooLong)
+    return result;
+
+  let cutQuotations = cutQuotation(xmlDocument);
+  // If message is too big try to rely on gmail quote to find checkpoint
+  if (result.isTooLong) {
+  const cutQuotationResult = extractQuotationFromCheckpoint(xmlDocument, messageBody, options);
```
Comment (quentez) [inline src/Quotations.ts]: Is there a point in trying again here if the `cutQuotation` process failed?

File: src/Quotations.ts [hunk 3]
```diff
@@ -111,49 +105,66 @@ export function extractFromHtml(messageBody: string, options?: ExtractFromHtmlOp
   if (xmlDocument.lastChild && xmlDocument.lastChild.nodeValue)
     xmlDocument.removeChild(xmlDocument.lastChild);
 
-  // Keep a copy of the original document around.
+  // Add the checkpoints to the HTML tree.
+  const result = extractQuotationFromCheckpoint(xmlDocument, messageBody, options);
+  if (result.didFindQuote && !result.isTooLong)
+    return result;
+
+  let cutQuotations = cutQuotation(xmlDocument);
```
Comment (quentez) [inline src/Quotations.ts]: Let's add a comment on what that does.

File: src/Quotations.ts (hunk 1 above)
Comment (quentez) [reply src/Quotations.ts]: One option: `attempt1` and `attempt2`. To make it clear that we're trying something similar multiple times.
But I don't feel strongly, we can keep `result`.

File: src/Quotations.ts [hunk 4]
```diff
@@ -111,49 +105,66 @@ export function extractFromHtml(messageBody: string, options?: ExtractFromHtmlOp
...
+    return {body: null, didFindQuote: false, isTooLong: false};
 
-    // Serialize and return.
-    return {
-      body: xmlDomSerializer.serializeToString(xmlDocumentCopy, true),
-      didFindQuote: true
-    }
-  }
+  // Remove the tags that we marked as quotation from the HTML.
+  deleteQuotationTags(xmlDocumentCopy, xmlDocument, extractQuoteHtml.quotationCheckpoints, {nodeLimit});
```
Comment (quentez) [inline src/Quotations.ts:156]: The logic was a bit different before where we were performing the detection on the `xmlDocument` and performing the deletion on the copy. I'm not sure I understand why we're inverting that pattern here.

File: src/Quotations.ts (hunk 2 above)
Comment (evonck) [reply src/Quotations.ts]: No good catch

File: src/Quotations.ts (hunk 4 above)
Comment (evonck) [reply src/Quotations.ts:156]: We used to add the marker on the xmlDocument and removing it from there. As we know might process the document twice (before removing quote and after), I change the logic to add the checkpoint on a copy of the xmlDocument (hence remove it from there). This allow us to reuse the xmlDocument after the gmail quote are removed.

**Review by quentez — APPROVED** [review]

**Pull request merged**